### PR TITLE
ClicksAndViews Example App

### PR DESCRIPTION
**Pull request by anwar6953**
Implementation of ClicksAndViews example application, which demonstrates a join across two streams, using multiple inputs functionality of MapReduce.

Note: Opening this PR to get feedback on the example application, while I work on the developer docs for it.

Passes [Quick Build 18](http://builds.cask.co/browse/CDAP-DOB184-18)

### CODE REVIEW

File: cdap-examples/ClicksAndViews/src/main/java/co/cask/cdap/examples/clicksandviews/ClicksAndViews.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    // Process the "clicks" and "views" streams using MapReduce
+    addMapReduce(new ClicksAndViewsMapReduce());
+
+    addStream(CLICKS);
+    addStream(VIEWS);
+
+    createDataset(JOINED, FileSet.class, FileSetProperties.builder()
+      // Properties for file set to be used as output of MapReduce
+      .setOutputFormat(TextOutputFormat.class)
```
Comment (gokulavasan) [inline cdap-examples/ClicksAndViews/src/main/java/co/cask/cdap/examples/clicksandviews/ClicksAndViews.java]: Can you add a description for the dataset?

File: cdap-examples/ClicksAndViews/src/main/java/co/cask/cdap/examples/clicksandviews/ClicksAndViews.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    setDescription("Example Clicks and Views Application");
+
+    // Process the "clicks" and "views" streams using MapReduce
+    addMapReduce(new ClicksAndViewsMapReduce());
+
+    addStream(new Stream(VIEWS, "Stores the views that happen for each ad"));
+    addStream(new Stream(CLICKS, "Stores the clicks that happen for each view"));
+
+    createDataset(JOINED, PartitionedFileSet.class, PartitionedFileSetProperties.builder()
+      // partition on `runtime`, represents the MapReduce's logical runtime
```
Comment (bdmogal) [inline cdap-examples/ClicksAndViews/src/main/java/co/cask/cdap/examples/clicksandviews/ClicksAndViews.java]: nit: use quotes instead of backticks?

File: cdap-examples/ClicksAndViews/src/main/java/co/cask/cdap/examples/clicksandviews/ClicksAndViews.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import co.cask.cdap.api.dataset.lib.PartitionedFileSet;
+import co.cask.cdap.api.dataset.lib.PartitionedFileSetProperties;
+import co.cask.cdap.api.dataset.lib.Partitioning;
+import org.apache.hadoop.mapreduce.lib.output.TextOutputFormat;
+
+/**
+ * Clicks and Views sample Application.
+ */
+public class ClicksAndViews extends AbstractApplication {
+  protected static final String NAME = "ClicksAndViews";
```
Comment (bdmogal) [inline cdap-examples/ClicksAndViews/src/main/java/co/cask/cdap/examples/clicksandviews/ClicksAndViews.java]: why are these protected? either package-private or public (if you want to use them outside the CDAP repo)?

File: cdap-examples/ClicksAndViews/src/test/java/co/cask/cdap/examples/clicksandviews/ClicksAndViewsMapReduceTest.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+                     createView(1, 1461265001, 2157, "http://www.google.co.uk", "lu=8fsdggknea@ASJHlz", "21.612.39.63"),
+                     createView(2, 1461281958, 2157, "http://www.yahoo.com", "name=Mike", "212.193.252.52"),
+                     createView(3, 1461331879, 2157, "http://www.amazon.com", "name=Matt", "1.116.135.146"),
+                     createView(4, 1461348738, 2157, "http://www.t.co", "name=Nicholas; Httponly", "89.141.94.158"),
+                     createView(5, 1461349158, 2157, "http://www.linkedin.com", "lo=Npa0jbIHGloMnx75", "69.75.87.114"));
+
+  private static final ImmutableList<Integer> CLICKS = ImmutableList.of(1, 2, 5);
+
+  private static final ImmutableSet<String> JOINED_VIEWS;
+  static {
```
Comment (bdmogal) [inline cdap-examples/ClicksAndViews/src/test/java/co/cask/cdap/examples/clicksandviews/ClicksAndViewsMapReduceTest.java]: don't need a static block right?  If it cannot be instantiated in place, just use ``@BeforeSubmit``?

File: cdap-examples/ClicksAndViews/src/main/java/co/cask/cdap/examples/clicksandviews/ClicksAndViewsMapReduce.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+import org.apache.hadoop.mapreduce.Mapper;
+import org.apache.hadoop.mapreduce.Reducer;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A MapReduce program that reads from a CLICKS stream and a VIEWS stream. It performs a join across these two streams,
+ * based upon the impressionId of the records. It then keeps track of how many clicks a particular view resulted in.
```
Comment (sagarkapare) [inline cdap-examples/ClicksAndViews/src/main/java/co/cask/cdap/examples/clicksandviews/ClicksAndViewsMapReduce.java]: Is `impressionId` same as `viewId`? did not see any explanation for it in rst.

Comment (anwar6953) [inline cdap-examples/ClicksAndViews/src/main/java/co/cask/cdap/examples/clicksandviews/ClicksAndViewsMapReduce.java]: Sorry, yes. I will make it consistent, changing 'impressionId' to 'viewId'.

File: cdap-examples/ClicksAndViews/src/main/java/co/cask/cdap/examples/clicksandviews/ClicksAndViews.java (hunk 2 above)
Comment (anwar6953) [inline cdap-examples/ClicksAndViews/src/main/java/co/cask/cdap/examples/clicksandviews/ClicksAndViews.java]: done. that also makes it consistent.

File: cdap-examples/ClicksAndViews/src/main/java/co/cask/cdap/examples/clicksandviews/ClicksAndViews.java (hunk 3 above)
Comment (anwar6953) [inline cdap-examples/ClicksAndViews/src/main/java/co/cask/cdap/examples/clicksandviews/ClicksAndViews.java]: made package-private.

File: cdap-examples/ClicksAndViews/src/test/java/co/cask/cdap/examples/clicksandviews/ClicksAndViewsMapReduceTest.java (hunk 4 above)
Comment (anwar6953) [inline cdap-examples/ClicksAndViews/src/test/java/co/cask/cdap/examples/clicksandviews/ClicksAndViewsMapReduceTest.java]: computing it in the test case itself now.

**Comment by bdmogal** [comment]
LGTM for my comments

File: cdap-docs/examples-manual/source/examples/clicks-and-views.rst [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+As in the other :ref:`examples <examples-index>`, the components
+of the application are tied together by the class ``ClicksAndViews``:
+
+.. literalinclude:: /../../../cdap-examples/ClicksAndViews/src/main/java/co/cask/cdap/examples/clicksandviews/ClicksAndViews.java
+   :language: java
+   :lines: 29-
+
+Data Storage
+------------
+- *views* input``Stream`` contains ad views, representing an advertisement downloaded to a viewer's screen.
```
Comment (anew) [inline cdap-docs/examples-manual/source/examples/clicks-and-views.rst]: may be better say, "displayed on a viewer's screen"?

File: cdap-docs/examples-manual/source/examples/clicks-and-views.rst [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+Ingesting Records
+-----------------
+Begin by uploading a file containing some newline-separated records representing view events into the *views* stream:
+
+.. tabbed-parsed-literal::
+
+  $ cdap-cli.sh load stream views examples/ClicksAndViews/resources/views.txt
+  
+  Successfully sent stream event to stream 'views'
```
Comment (anew) [inline cdap-docs/examples-manual/source/examples/clicks-and-views.rst]: Is this really the response message from CLI? It seems wrong, it has actually sent a file full of events.

File: cdap-docs/examples-manual/source/examples/clicks-and-views.rst (hunk 6 above)
Comment (anwar6953) [inline cdap-docs/examples-manual/source/examples/clicks-and-views.rst]: updated.

File: cdap-docs/examples-manual/source/examples/clicks-and-views.rst [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+  count of clicks each ad view has. This Dataset is partitioned on the logical start time of the MapReduce.
+
+MapReduce over Multiple Inputs
+------------------------------
+``ClicksAndViewsMapReduce`` is a MapReduce that reads from the *clicks* and *views* ``Stream``s and writes to
+the *joined* ``PartitionedFileSet``. The ``beforeSubmit`` method prepares the MapReduce program. It sets up the
+two ``Stream``s as input and sets up the ``PartitionedFileSet`` as output, with the appropriate ``PartitionKey``.
+The Mapper class then keys each of the records based upon the ``viewId``. That makes it so that all clicks and views
+for a particular ``viewId`` go to a single Reducer for joining.
+
```
Comment (anew) [inline cdap-docs/examples-manual/source/examples/clicks-and-views.rst]: It would be good to include the code snippets from the MapReduce that actually deal with the multiple inputs, and explain what they do and how.

File: cdap-examples/ClicksAndViews/src/main/java/co/cask/cdap/examples/clicksandviews/ClicksAndViewsMapReduce.java [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+      Preconditions.checkArgument(ClicksAndViews.CLICKS.equals(inputName) || ClicksAndViews.VIEWS.equals(inputName));
+    }
+
+    @Override
+    public void destroy() {
+    }
+
+    @Override
+    protected void map(LongWritable key, Text value, Context context) throws IOException, InterruptedException {
+      String[] split = value.toString().split("\t");
```
Comment (anew) [inline cdap-examples/ClicksAndViews/src/main/java/co/cask/cdap/examples/clicksandviews/ClicksAndViewsMapReduce.java]: a little misleading to call this split. (A split is a different thing in MapReduce). Perhaps call it fields, or fragments?

Comment (anwar6953) [inline cdap-examples/ClicksAndViews/src/main/java/co/cask/cdap/examples/clicksandviews/ClicksAndViewsMapReduce.java]: will rename it `parts`.

**Comment by anew** [comment]
I did a pass over this, please see comments.

File: cdap-docs/examples-manual/source/examples/clicks-and-views.rst (hunk 8 above)
Comment (anwar6953) [inline cdap-docs/examples-manual/source/examples/clicks-and-views.rst]: I added the code snippet pointing to the beforeSubmit (the part of setting input and output).
What did you mean by "explain what they do and how"

File: cdap-docs/examples-manual/source/examples/clicks-and-views.rst (hunk 7 above)
Comment (anwar6953) [inline cdap-docs/examples-manual/source/examples/clicks-and-views.rst]: That's what is printed by LoadStreamCommand. I agree though, it seems like an improper message.
How about "Successfully loaded file to stream 'views'"?

**Comment by anwar6953** [comment]
@anew addressed comments; thanks.

File: cdap-docs/examples-manual/source/examples/clicks-and-views.rst (hunk 7 above)
Comment (anew) [inline cdap-docs/examples-manual/source/examples/clicks-and-views.rst]: Yes, but we should fix  that separately

File: cdap-docs/examples-manual/source/examples/clicks-and-views.rst (hunk 8 above)
Comment (anew) [inline cdap-docs/examples-manual/source/examples/clicks-and-views.rst]: I meant that what you had was only a description of _what_ the MR does, in a couple of lines of text, without showing the code that does it. The purpose of the example, however, is to illustrate how to use the APIs to make a MR do that. So it would be better to include code samples, and for each one, explain what it is doing.

File: cdap-docs/examples-manual/source/examples/clicks-and-views.rst [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+MapReduce over Multiple Inputs
+------------------------------
+``ClicksAndViewsMapReduce`` is a MapReduce that reads from the *clicks* and *views* ``Stream``s and writes to
+the *joined* ``PartitionedFileSet``. The ``beforeSubmit`` method prepares the MapReduce program. It sets up the
+two ``Stream``s as input and sets up the ``PartitionedFileSet`` as output, with the appropriate ``PartitionKey``:
+
+.. literalinclude:: /../../../cdap-examples/ClicksAndViews/src/main/java/co/cask/cdap/examples/clicksandviews/ClicksAndViewsMapReduce.java
+:language: java
+     :lines: 58-73
```
Comment (anew) [inline cdap-docs/examples-manual/source/examples/clicks-and-views.rst]: this is only the beforeSubmit(). In the mapper itself, you also use a feature that is worth explaining - context.getInputName(). I think you should include the mapper code here.

**Comment by anew** [comment]
I did another pass. Please include the mapper in the documentation, that is my only remaining comment.

File: cdap-docs/examples-manual/source/examples/clicks-and-views.rst [hunk 11]
```diff
@@ -44,11 +44,18 @@ the *joined* ``PartitionedFileSet``. The ``beforeSubmit`` method prepares the Ma
...
 .. literalinclude:: /../../../cdap-examples/ClicksAndViews/src/main/java/co/cask/cdap/examples/clicksandviews/ClicksAndViewsMapReduce.java
-:language: java
-     :lines: 58-73
+   :language: java
+   :lines: 58-73
 
 The Mapper class then keys each of the records based upon the ``viewId``. That makes it so that all clicks and views
-for a particular ``viewId`` go to a single Reducer for joining.
+for a particular ``viewId`` go to a single Reducer for joining. In order to do the join properly, the Reducer needs
+to know which source each record is coming from. To make this possible, the Mapper class can get access to the source
```
Comment (anew) [inline cdap-docs/examples-manual/source/examples/clicks-and-views.rst]: source/input

Comment (anew) [inline cdap-docs/examples-manual/source/examples/clicks-and-views.rst]: I think this sentence is hard to parse. Better: 
This is possible by calling the ``getInputName`` method of the ``MapReduceTaskContext`` that is passed to ``initialize()``. Note that the Mapper needs to implement ProgramLifeCycle:

**Comment by anew** [comment]
Two more comments to make the doc a little easier to read. Other than that LGTM.

**Comment by awholegunch** [comment]
Docs LGTM. I made two commits. please check.

**Comment by anwar6953** [comment]
@awholegunch's changes lgtm.
@anew any other comments?

**Comment by anwar6953** [comment]
Squashed commits.

**Comment by anwar6953** [comment]
Build is green:
http://builds.cask.co/browse/CDAP-DUT4017-15
